File: java/lib/src/main/java/com/svix/api/Connector.java
// this file is @generated
package com.svix.api;

import com.svix.SvixHttpClient;
import com.svix.Utils;
import com.svix.exceptions.ApiException;
import com.svix.models.ConnectorIn;
import com.svix.models.ConnectorOut;
import com.svix.models.ConnectorPatch;
import com.svix.models.ConnectorUpdate;
import com.svix.models.ListResponseConnectorOut;

import okhttp3.Headers;
import okhttp3.HttpUrl;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public class Connector {
    private final SvixHttpClient client;

    public Connector(SvixHttpClient client) {
        this.client = client;
    }

    /** List all connectors for an application. */
    public ListResponseConnectorOut list() throws IOException, ApiException {

        return this.list(new ConnectorListOptions());
    }

    /** List all connectors for an application. */
    public ListResponseConnectorOut list(final ConnectorListOptions options)
            throws IOException, ApiException {
        HttpUrl.Builder url = this.client.newUrlBuilder().encodedPath("/api/v1/connector");
        if (options.limit != null) {
            url.addQueryParameter("limit", Utils.serializeQueryParam(options.limit));
        }
        if (options.iterator != null) {
            url.addQueryParameter("iterator", options.iterator);
        }
        if (options.order != null) {
            url.addQueryParameter("order", Utils.serializeQueryParam(options.order));
        }
        if (options.productType != null) {
            url.addQueryParameter("product_type", Utils.serializeQueryParam(options.productType));
        }
        return this.client.executeRequest(
                "GET", url.build(), null, null, ListResponseConnectorOut.class);
    }

    /** Create a new connector. */
    public ConnectorOut create(final ConnectorIn connectorIn) throws IOException, ApiException {
        return this.create(connectorIn, new ConnectorCreateOptions());
    }

    /** Create a new connector. */
    public ConnectorOut create(final ConnectorIn connectorIn, final ConnectorCreateOptions options)
            throws IOException, ApiException {
        HttpUrl.Builder url = this.client.newUrlBuilder().encodedPath("/api/v1/connector");
        Map<String, String> headers = new HashMap<>();
        if (options.idempotencyKey != null) {
            headers.put("idempotency-key", options.idempotencyKey);
        }
        return this.client.executeRequest(
                "POST", url.build(), Headers.of(headers), connectorIn, ConnectorOut.class);
    }

    /** Get a connector. */
    public ConnectorOut get(final String connectorId) throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(String.format("/api/v1/connector/%s", connectorId));
        return this.client.executeRequest("GET", url.build(), null, null, ConnectorOut.class);
    }

    /** Update a connector. */
    public ConnectorOut update(final String connectorId, final ConnectorUpdate connectorUpdate)
            throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(String.format("/api/v1/connector/%s", connectorId));
        return this.client.executeRequest(
                "PUT", url.build(), null, connectorUpdate, ConnectorOut.class);
    }

    /** Delete a connector. */
    public void delete(final String connectorId) throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(String.format("/api/v1/connector/%s", connectorId));
        this.client.executeRequest("DELETE", url.build(), null, null, null);
    }

    /** Partially update a connector. */
    public ConnectorOut patch(final String connectorId, final ConnectorPatch connectorPatch)
            throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(String.format("/api/v1/connector/%s", connectorId));
        return this.client.executeRequest(
                "PATCH", url.build(), null, connectorPatch, ConnectorOut.class);
    }
}

File: kotlin/lib/src/main/kotlin/Connector.kt
// this file is @generated
package com.svix.kotlin

import com.svix.kotlin.models.ConnectorIn
import com.svix.kotlin.models.ConnectorOut
import com.svix.kotlin.models.ConnectorPatch
import com.svix.kotlin.models.ConnectorUpdate
import com.svix.kotlin.models.ListResponseConnectorOut
import com.svix.kotlin.models.Ordering
import okhttp3.Headers

data class ConnectorListOptions(
    /** Limit the number of returned items */
    val limit: ULong? = null,
    /** The iterator returned from a prior invocation */
    val iterator: String? = null,
    /** The sorting order of the returned items */
    val order: Ordering? = null,
)

data class ConnectorCreateOptions(val idempotencyKey: String? = null)

class Connector(private val client: SvixHttpClient) {
    /** List all connectors for an application. */
    suspend fun list(
        options: ConnectorListOptions = ConnectorListOptions()
    ): ListResponseConnectorOut {
        val url = client.newUrlBuilder().encodedPath("/api/v1/connector")
        options.limit?.let { url.addQueryParameter("limit", serializeQueryParam(it)) }
        options.iterator?.let { url.addQueryParameter("iterator", it) }
        options.order?.let { url.addQueryParameter("order", serializeQueryParam(it)) }
        return client.executeRequest<Any, ListResponseConnectorOut>("GET", url.build())
    }

    /** Create a new connector. */
    suspend fun create(
        connectorIn: ConnectorIn,
        options: ConnectorCreateOptions = ConnectorCreateOptions(),
    ): ConnectorOut {
        val url = client.newUrlBuilder().encodedPath("/api/v1/connector")
        val headers = Headers.Builder()
        options.idempotencyKey?.let { headers.add("idempotency-key", it) }

        return client.executeRequest<ConnectorIn, ConnectorOut>(
            "POST",
            url.build(),
            headers = headers.build(),
            reqBody = connectorIn,
        )
    }

    /** Get a connector. */
    suspend fun get(connectorId: String): ConnectorOut {
        val url = client.newUrlBuilder().encodedPath("/api/v1/connector/$connectorId")
        return client.executeRequest<Any, ConnectorOut>("GET", url.build())
    }

    /** Update a connector. */
    suspend fun update(connectorId: String, connectorUpdate: ConnectorUpdate): ConnectorOut {
        val url = client.newUrlBuilder().encodedPath("/api/v1/connector/$connectorId")

        return client.executeRequest<ConnectorUpdate, ConnectorOut>(
            "PUT",
            url.build(),
            reqBody = connectorUpdate,
        )
    }

    /** Delete a connector. */
    suspend fun delete(connectorId: String) {
        val url = client.newUrlBuilder().encodedPath("/api/v1/connector/$connectorId")
        client.executeRequest<Any, Boolean>("DELETE", url.build())
    }

    /** Partially update a connector. */
    suspend fun patch(connectorId: String, connectorPatch: ConnectorPatch): ConnectorOut {
        val url = client.newUrlBuilder().encodedPath("/api/v1/connector/$connectorId")

        return client.executeRequest<ConnectorPatch, ConnectorOut>(
            "PATCH",
            url.build(),
            reqBody = connectorPatch,
        )
    }
}
